lib/cylc/network/client.py

```diff
@@ -0,0 +1,296 @@
+        timeout (int):
+            Message receive timeout in seconds. Also used to set the
+            "linger" time, see ``ZMQClient``.
+
+    Determine host and port from the contact file unless provided.
+
+    If there is no socket bound to the specified host/port the client will
+    bail after ``timeout`` seconds.
+
+    """
+
+    NOT_RUNNING = "Contact info not found for suite \"%s\", suite not running?"
+
+    def __new__(cls, suite, owner=None, host=None, port=None, timeout=None):
+        if isinstance(timeout, str):
+            timeout = float(timeout)
+
+        # work out what we are connecting to
+        if port:
+            port = int(port)
+        if not (host and port):
+            host, port = cls.get_location(suite, owner, host)
+
+        # create connection
+        return ZMQClient(
+            host, port, encrypt, decrypt, lambda: get_secret(suite),
+            timeout=timeout, header=cls.get_header(),
+            timeout_handler=lambda: cls._timeout_handler(suite, host, port)
+        )
+
+    @staticmethod
+    def get_header():
+        """Return "header" data to attach to each request for traceability."""
+        CYLC_EXE = os.path.join(os.environ['CYLC_DIR'], 'bin', '')
+        cmd = sys.argv[0]
+
+        if cmd.startswith(CYLC_EXE):
+            cmd = cmd.replace(CYLC_EXE, '')
+
+        return {
+            'meta': {
+                'prog': cmd,
+                'host': socket.gethostname()
+            }
+        }
+
+    @staticmethod
+    def _timeout_handler(suite, host, port):
+        """Handle the eventuality of a communication timeout with the suite."""
+        if suite is None:
+            return
+        # Cannot connect, perhaps suite is no longer running and is leaving
+        # behind a contact file?
+        try:
+            SuiteSrvFilesManager().detect_old_contact_file(suite, (host, port))
+        except (AssertionError, SuiteServiceFileError):
+            # * contact file not have matching (host, port) to suite proc
+            # * old contact file exists and the suite process still alive
+            return
+        else:
+            # the suite has stopped
+            raise ClientError(  # TODO: SuiteStoppedError?
+                {'message': 'Suite "%s" already stopped' % suite})
+
+    @classmethod
+    def get_location(cls, suite, owner, host):
+        """Extract host and port from a suite's contact file."""
+        try:
+            contact = SuiteSrvFilesManager().load_contact_file(
+                suite, owner, host)
+        except SuiteServiceFileError:
+            sys.exit(cls.NOT_RUNNING % suite)
+            # monkey-patch the error message to make it more informative.
+            # exc.args = (cls.NOT_RUNNING % suite,)
+            # raise
+
+        if host and host.split('.')[0] == 'localhost':
+            host = get_host()
+        elif host and '.' not in host:  # Not IP and no domain
+            host = get_fqdn_by_host(host)
```

Comment:
Should use `cylc.hostuserutil` module to these stuffs?

Reply:
Could move it there.

Reply:
Can do.